File: go/spore-core/promptchunkregistry/promptchunkregistry_test.go
package promptchunkregistry

import (
	"errors"
	"strings"
	"testing"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

func registryWithRole(t *testing.T, id ChunkID) *StandardPromptChunkRegistry {
	t.Helper()
	r := NewStandardPromptChunkRegistry()
	if err := r.Register(NewPromptChunk(id, "you are a test agent", ChunkSlotRole, CacheBlockStatic)); err != nil {
		t.Fatalf("seed register: %v", err)
	}
	return r
}

// ── Register: error cases ───────────────────────────────────────────────────

func TestRegister_ErrorCases(t *testing.T) {
	cases := []struct {
		name      string
		setup     func(r *StandardPromptChunkRegistry)
		chunk     PromptChunk
		assertErr func(t *testing.T, err error)
	}{
		{
			name: "duplicate_id",
			setup: func(r *StandardPromptChunkRegistry) {
				_ = r.Register(NewPromptChunk("x", "hello", ChunkSlotCapability, CacheBlockStatic))
			},
			chunk: NewPromptChunk("x", "world", ChunkSlotCapability, CacheBlockStatic),
			assertErr: func(t *testing.T, err error) {
				var dup *ErrDuplicateID
				if !errors.As(err, &dup) {
					t.Fatalf("expected *ErrDuplicateID, got %T: %v", err, err)
				}
				if dup.ID != "x" {
					t.Errorf("want id=x, got %q", dup.ID)
				}
			},
		},
		{
			name:  "empty_content",
			chunk: NewPromptChunk("x", "   ", ChunkSlotCapability, CacheBlockStatic),
			assertErr: func(t *testing.T, err error) {
				var bad *InvalidSlotError
				if !errors.As(err, &bad) {
					t.Fatalf("expected *InvalidSlotError, got %T: %v", err, err)
				}
			},
		},
		{
			name:  "budget_slot_rejects_static",
			chunk: NewPromptChunk("b", "budget warning", ChunkSlotBudget, CacheBlockStatic),
			assertErr: func(t *testing.T, err error) {
				var c *ConflictingCacheBlockError
				if !errors.As(err, &c) {
					t.Fatalf("expected *ConflictingCacheBlockError, got %T: %v", err, err)
				}
				if c.Slot != ChunkSlotBudget || c.Expected != CacheBlockPerTurn || c.Actual != CacheBlockStatic {
					t.Errorf("unexpected fields: %+v", c)
				}
			},
		},
		{
			name:  "ephemeral_slot_rejects_per_session",
			chunk: NewPromptChunk("e", "ephemeral", ChunkSlotEphemeral, CacheBlockPerSession),
			assertErr: func(t *testing.T, err error) {
				var c *ConflictingCacheBlockError
				if !errors.As(err, &c) {
					t.Fatalf("expected *ConflictingCacheBlockError, got %T: %v", err, err)
				}
			},
		},
		{
			name:  "role_slot_rejects_per_session",
			chunk: NewPromptChunk("r", "role", ChunkSlotRole, CacheBlockPerSession),
			assertErr: func(t *testing.T, err error) {
				var c *ConflictingCacheBlockError
				if !errors.As(err, &c) {
					t.Fatalf("expected *ConflictingCacheBlockError, got %T: %v", err, err)
				}
			},
		},
		{
			name:  "mode_slot_rejects_per_turn",
			chunk: NewPromptChunk("m", "mode", ChunkSlotMode, CacheBlockPerTurn),
			assertErr: func(t *testing.T, err error) {
				var c *ConflictingCacheBlockError
				if !errors.As(err, &c) {
					t.Fatalf("expected *ConflictingCacheBlockError, got %T: %v", err, err)
				}
			},
		},
	}

	for _, tc := range cases {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			r := NewStandardPromptChunkRegistry()
			if tc.setup != nil {
				tc.setup(r)
			}
			err := r.Register(tc.chunk)
			if err == nil {
				t.Fatal("expected error, got nil")
			}
			tc.assertErr(t, err)
		})
	}
}

// ── Compose ─────────────────────────────────────────────────────────────────

func TestCompose_MissingRoleReturnsError(t *testing.T) {
	r := NewStandardPromptChunkRegistry()
	_, err := r.Compose("missing", ModeYolo, nil, nil)
	if err == nil {
		t.Fatal("expected error")
	}
	verrs := AsValidationErrors(err)
	found := false
	for _, e := range verrs {
		var m *MissingRequiredSlotError
		if errors.As(e, &m) && m.Slot == ChunkSlotRole {
			found = true
		}
	}
	if !found {
		t.Fatalf("expected MissingRequiredSlot{Role} in %v", verrs)
	}
}

func TestCompose_IncludesModeChunkFromEnum(t *testing.T) {
	r := registryWithRole(t, "role-test")
	composed, err := r.Compose("role-test", ModePlan, nil, nil)
	if err != nil {
		t.Fatalf("compose: %v", err)
	}
	var modeID ChunkID
	for _, c := range composed.Chunks {
		if c.Slot == ChunkSlotMode {
			modeID = c.ID
		}
	}
	if modeID != "mode-plan" {
		t.Fatalf("expected mode-plan, got %q", modeID)
	}
}

func TestCompose_OrdersBySlot(t *testing.T) {
	r := registryWithRole(t, "role-test")
	if err := r.Register(NewPromptChunk("cap-1", "cap one", ChunkSlotCapability, CacheBlockStatic)); err != nil {
		t.Fatal(err)
	}
	if err := r.Register(NewPromptChunk("skill-1", "skill one", ChunkSlotSkill, CacheBlockStatic)); err != nil {
		t.Fatal(err)
	}
	composed, err := r.Compose("role-test", ModeAutoEdit, []ChunkID{"cap-1"}, []ChunkID{"skill-1"})
	if err != nil {
		t.Fatalf("compose: %v", err)
	}
	want := []ChunkSlot{ChunkSlotRole, ChunkSlotMode, ChunkSlotCapability, ChunkSlotSkill}
	if len(composed.Chunks) != len(want) {
		t.Fatalf("len mismatch: got %d, want %d", len(composed.Chunks), len(want))
	}
	for i, w := range want {
		if composed.Chunks[i].Slot != w {
			t.Errorf("chunks[%d].Slot = %q, want %q", i, composed.Chunks[i].Slot, w)
		}
	}
}

func TestCompose_PreservesArgOrderWithinSlot(t *testing.T) {
	r := registryWithRole(t, "role-test")
	for _, id := range []ChunkID{"cap-a", "cap-b", "cap-c"} {
		if err := r.Register(NewPromptChunk(id, string(id), ChunkSlotCapability, CacheBlockStatic)); err != nil {
			t.Fatal(err)
		}
	}
	composed, err := r.Compose("role-test", ModeYolo, []ChunkID{"cap-c", "cap-a", "cap-b"}, nil)
	if err != nil {
		t.Fatalf("compose: %v", err)
	}
	var caps []ChunkID
	for _, c := range composed.Chunks {
		if c.Slot == ChunkSlotCapability {
			caps = append(caps, c.ID)
		}
	}
	want := []ChunkID{"cap-c", "cap-a", "cap-b"}
	for i, w := range want {
		if caps[i] != w {
			t.Errorf("caps[%d]=%q, want %q", i, caps[i], w)
		}
	}
}

// ── Block hashes ────────────────────────────────────────────────────────────

func TestBlockHashes_StableForIdenticalContent(t *testing.T) {
	a, err := registryWithRole(t, "role-test").Compose("role-test", ModeYolo, nil, nil)
	if err != nil {
		t.Fatal(err)
	}
	b, err := registryWithRole(t, "role-test").Compose("role-test", ModeYolo, nil, nil)
	if err != nil {
		t.Fatal(err)
	}
	if a.Block1Hash != b.Block1Hash {
		t.Errorf("block1 hashes differ: %d vs %d", a.Block1Hash, b.Block1Hash)
	}
	if a.Block2Hash != b.Block2Hash {
		t.Errorf("block2 hashes differ: %d vs %d", a.Block2Hash, b.Block2Hash)
	}
}

func TestBlock1Hash_ChangesWhenContentChanges(t *testing.T) {
	a, err := registryWithRole(t, "role-test").Compose("role-test", ModeYolo, nil, nil)
	if err != nil {
		t.Fatal(err)
	}
	r2 := NewStandardPromptChunkRegistry()
	if err := r2.Register(NewPromptChunk("role-test", "DIFFERENT ROLE CONTENT", ChunkSlotRole, CacheBlockStatic)); err != nil {
		t.Fatal(err)
	}
	b, err := r2.Compose("role-test", ModeYolo, nil, nil)
	if err != nil {
		t.Fatal(err)
	}
	if a.Block1Hash == b.Block1Hash {
		t.Error("block1 hash should change when role content changes")
	}
}

// ── Validate ────────────────────────────────────────────────────────────────

func TestValidate_FlagsPerTurnChunkInStaticBlock(t *testing.T) {
	r := NewStandardPromptChunkRegistry()
	composed := &ComposedPrompt{
		Chunks: []PromptChunk{
			NewPromptChunk("role-x", "x", ChunkSlotRole, CacheBlockStatic),
			ModeYolo.PromptChunk(),
			// Budget chunk with Static cache block — simulates a bug.
			{ID: "bad-budget", Content: "b", Slot: ChunkSlotBudget, CacheBlock: CacheBlockStatic},
		},
	}
	errs := r.Validate(composed)
	found := false
	for _, e := range errs {
		var pe *PerTurnChunkInStaticBlockError
		if errors.As(e, &pe) && pe.ID == "bad-budget" {
			found = true
		}
	}
	if !found {
		t.Fatalf("expected PerTurnChunkInStaticBlock for bad-budget, got %v", errs)
	}
}

func TestValidate_FlagsMoreThanOneModeChunk(t *testing.T) {
	r := NewStandardPromptChunkRegistry()
	composed := &ComposedPrompt{
		Chunks: []PromptChunk{
			NewPromptChunk("role-x", "x", ChunkSlotRole, CacheBlockStatic),
			ModeYolo.PromptChunk(),
			ModeAlwaysAsk.PromptChunk(),
		},
	}
	errs := r.Validate(composed)
	found := false
	for _, e := range errs {
		var c *ConflictingModeChunksError
		if errors.As(e, &c) {
			found = true
		}
	}
	if !found {
		t.Fatalf("expected ConflictingModeChunks, got %v", errs)
	}
}

func TestValidate_FlagsMissingRoleSlot(t *testing.T) {
	r := NewStandardPromptChunkRegistry()
	composed := &ComposedPrompt{
		Chunks: []PromptChunk{ModeYolo.PromptChunk()},
	}
	errs := r.Validate(composed)
	found := false
	for _, e := range errs {
		var m *MissingRequiredSlotError
		if errors.As(e, &m) && m.Slot == ChunkSlotRole {
			found = true
		}
	}
	if !found {
		t.Fatalf("expected MissingRequiredSlot{Role}, got %v", errs)
	}
}

// ── Get ─────────────────────────────────────────────────────────────────────

func TestGet_ReturnsRegisteredChunk(t *testing.T) {
	r := registryWithRole(t, "role-x")
	c, ok := r.Get("role-x")
	if !ok {
		t.Fatal("expected found")
	}
	if c.ID != "role-x" {
		t.Errorf("id mismatch: %q", c.ID)
	}
	if _, ok := r.Get("nope"); ok {
		t.Error("expected not found")
	}
}

// ── Mode helpers ────────────────────────────────────────────────────────────

func TestMode_ApprovalPolicyMatchesSpec(t *testing.T) {
	cases := []struct {
		m    Mode
		want ApprovalPolicy
	}{
		{ModeAlwaysAsk, ApprovalPolicyAlwaysAsk},
		{ModeAutoEdit, ApprovalPolicyAutoExplain},
		{ModePlan, ApprovalPolicyPlanOnly},
		{ModeSafeAuto, ApprovalPolicySafeAuto},
		{ModeYolo, ApprovalPolicyNone},
	}
	for _, c := range cases {
		if got := c.m.ApprovalPolicy(); got != c.want {
			t.Errorf("%s: got %q, want %q", c.m, got, c.want)
		}
	}
}

func TestMode_DefaultToolPhase(t *testing.T) {
	if got := ModePlan.DefaultToolPhase(); got != sporecore.PhasePlanning {
		t.Errorf("Plan: %q", got)
	}
	for _, m := range []Mode{ModeAlwaysAsk, ModeAutoEdit, ModeSafeAuto, ModeYolo} {
		if got := m.DefaultToolPhase(); got != sporecore.PhaseExecution {
			t.Errorf("%s: %q", m, got)
		}
	}
}

func TestMode_PromptChunkIDsAndPrefixes(t *testing.T) {
	cases := []struct {
		m      Mode
		id     ChunkID
		prefix string
	}{
		{ModeAlwaysAsk, "mode-always-ask", "Mode: AlwaysAsk."},
		{ModeAutoEdit, "mode-auto-edit", "Mode: AutoEdit."},
		{ModePlan, "mode-plan", "Mode: Plan."},
		{ModeSafeAuto, "mode-safe-auto", "Mode: SafeAuto."},
		{ModeYolo, "mode-yolo", "Mode: Yolo."},
	}
	for _, c := range cases {
		pc := c.m.PromptChunk()
		if pc.ID != c.id {
			t.Errorf("%s: id=%q, want %q", c.m, pc.ID, c.id)
		}
		if !strings.HasPrefix(pc.Content, c.prefix) {
			t.Errorf("%s: content %q missing prefix %q", c.m, pc.Content, c.prefix)
		}
		if pc.Slot != ChunkSlotMode || pc.CacheBlock != CacheBlockStatic {
			t.Errorf("%s: slot=%q cache=%q", c.m, pc.Slot, pc.CacheBlock)
		}
	}
}

// ── ComposedPrompt rendering ────────────────────────────────────────────────

func TestComposedPrompt_RenderJoinsChunks(t *testing.T) {
	r := registryWithRole(t, "role-test")
	composed, err := r.Compose("role-test", ModeYolo, nil, nil)
	if err != nil {
		t.Fatal(err)
	}
	if composed.HasRendered() {
		t.Error("expected no cached render initially")
	}
	rendered := composed.Render()
	if !strings.Contains(rendered, "you are a test agent") {
		t.Error("missing role content")
	}
	if !strings.Contains(rendered, "Mode: Yolo") {
		t.Error("missing mode content")
	}
	if !composed.HasRendered() {
		t.Error("expected cached render after Render()")
	}
}

// ── Standard library smoke ──────────────────────────────────────────────────

func TestStandardChunks_RegisterCleanly(t *testing.T) {
	r := NewStandardPromptChunkRegistry()
	if err := r.RegisterStandardChunks(); err != nil {
		t.Fatalf("register standard: %v", err)
	}
	if _, ok := r.Get("role-coding-agent"); !ok {
		t.Error("missing role-coding-agent")
	}
	if _, ok := r.Get("capability-bash"); !ok {
		t.Error("missing capability-bash")
	}
	if _, ok := r.Get("skill-testing"); !ok {
		t.Error("missing skill-testing")
	}
}

func TestCompose_WithStandardChunks_CodingAgent(t *testing.T) {
	r := NewStandardPromptChunkRegistry()
	if err := r.RegisterStandardChunks(); err != nil {
		t.Fatal(err)
	}
	composed, err := r.Compose(
		"role-coding-agent",
		ModeSafeAuto,
		[]ChunkID{"capability-bash", "capability-filesystem", "capability-git"},
		[]ChunkID{"skill-testing", "skill-security-review"},
	)
	if err != nil {
		t.Fatal(err)
	}
	if want := 1 + 1 + 3 + 2; len(composed.Chunks) != want {
		t.Errorf("len=%d, want %d", len(composed.Chunks), want)
	}
	if composed.Chunks[0].Slot != ChunkSlotRole {
		t.Errorf("first slot = %q", composed.Chunks[0].Slot)
	}
	if composed.Chunks[1].Slot != ChunkSlotMode || composed.Chunks[1].ID != "mode-safe-auto" {
		t.Errorf("second chunk = %+v", composed.Chunks[1])
	}
}

File: typescript/packages/core/tests/prompt-chunk-registry.test.ts
/**
 * Unit tests for {@link PromptChunkRegistry} (spore-core issue #24).
 *
 * Covers every rule in the spec: registration validation, compose ordering
 * and missing-slot errors, block-hash stability, validate() rules, Mode
 * helpers, ComposedPrompt rendering, and the standard chunk library.
 */

import { describe, expect, it } from "vitest";

import { promptChunkRegistry as pcr } from "../src/index.js";

const {
  ChunkId,
  StandardPromptChunkRegistry,
  hashContent,
  modeApprovalPolicy,
  modeDefaultToolPhase,
  modePromptChunk,
  promptChunk,
  recomputeBlockHashes,
  renderComposed,
  renderedStr,
  standardChunks,
} = pcr;

function registryWithRole(id: string): InstanceType<typeof StandardPromptChunkRegistry> {
  const r = new StandardPromptChunkRegistry();
  expect(r.register(promptChunk(id, "you are a test agent", "role", "static"))).toBeNull();
  return r;
}

// ── Rule: register rejects duplicate ids ───────────────────────────────────

describe("register", () => {
  it("rejects duplicate ids", () => {
    const r = new StandardPromptChunkRegistry();
    expect(r.register(promptChunk("x", "hello", "capability", "static"))).toBeNull();
    const err = r.register(promptChunk("x", "world", "capability", "static"));
    expect(err?.kind).toBe("duplicate_id");
  });

  it("rejects empty content", () => {
    const r = new StandardPromptChunkRegistry();
    const err = r.register(promptChunk("x", "   ", "capability", "static"));
    expect(err?.kind).toBe("invalid_slot");
  });

  it("rejects Budget slot with Static cache block", () => {
    const r = new StandardPromptChunkRegistry();
    const err = r.register(promptChunk("b", "budget warning", "budget", "static"));
    expect(err?.kind).toBe("conflicting_cache_block");
    if (err && err.kind === "conflicting_cache_block") {
      expect(err.slot).toBe("budget");
      expect(err.expected).toBe("per_turn");
      expect(err.actual).toBe("static");
    }
  });

  it("rejects Ephemeral slot with PerSession cache block", () => {
    const r = new StandardPromptChunkRegistry();
    const err = r.register(promptChunk("e", "ephemeral", "ephemeral", "per_session"));
    expect(err?.kind).toBe("conflicting_cache_block");
  });

  it("rejects Role slot with non-Static cache block", () => {
    const r = new StandardPromptChunkRegistry();
    const err = r.register(promptChunk("r", "role", "role", "per_session"));
    expect(err?.kind).toBe("conflicting_cache_block");
  });

  it("rejects Mode slot with non-Static cache block", () => {
    const r = new StandardPromptChunkRegistry();
    const err = r.register(promptChunk("m", "mode", "mode", "per_turn"));
    expect(err?.kind).toBe("conflicting_cache_block");
  });
});

// ── Rule: compose ───────────────────────────────────────────────────────────

describe("compose", () => {
  it("returns missing-role error when role chunk is not registered", () => {
    const r = new StandardPromptChunkRegistry();
    const res = r.compose(new ChunkId("missing"), "yolo", [], []);
    expect(res.ok).toBe(false);
    if (!res.ok) {
      expect(res.errors.some((e) => e.kind === "missing_required_slot" && e.slot === "role")).toBe(
        true,
      );
    }
  });

  it("includes the mode chunk via modePromptChunk()", () => {
    const r = registryWithRole("role-test");
    const res = r.compose(new ChunkId("role-test"), "plan", [], []);
    expect(res.ok).toBe(true);
    if (res.ok) {
      const mode = res.composed.chunks.find((c) => c.slot === "mode");
      expect(mode?.id.value).toBe("mode-plan");
    }
  });

  it("orders chunks by slot (Role, Mode, Capability, Skill)", () => {
    const r = registryWithRole("role-test");
    expect(r.register(promptChunk("cap-1", "cap one", "capability", "static"))).toBeNull();
    expect(r.register(promptChunk("skill-1", "skill one", "skill", "static"))).toBeNull();
    const res = r.compose(
      new ChunkId("role-test"),
      "auto_edit",
      [new ChunkId("cap-1")],
      [new ChunkId("skill-1")],
    );
    expect(res.ok).toBe(true);
    if (res.ok) {
      expect(res.composed.chunks.map((c) => c.slot)).toEqual([
        "role",
        "mode",
        "capability",
        "skill",
      ]);
    }
  });

  it("returns missing-capability error when capability chunk is absent", () => {
    const r = registryWithRole("role-test");
    const res = r.compose(new ChunkId("role-test"), "yolo", [new ChunkId("nope")], []);
    expect(res.ok).toBe(false);
    if (!res.ok) {
      expect(
        res.errors.some((e) => e.kind === "missing_required_slot" && e.slot === "capability"),
      ).toBe(true);
    }
  });
});

// ── Rule: block hashes ──────────────────────────────────────────────────────

describe("block hashes", () => {
  it("are stable for identical content", () => {
    const a = registryWithRole("role-test").compose(new ChunkId("role-test"), "yolo", [], []);
    const b = registryWithRole("role-test").compose(new ChunkId("role-test"), "yolo", [], []);
    expect(a.ok && b.ok).toBe(true);
    if (a.ok && b.ok) {
      expect(a.composed.block_1_hash).toBe(b.composed.block_1_hash);
      expect(a.composed.block_2_hash).toBe(b.composed.block_2_hash);
    }
  });

  it("block_1_hash changes when Static content changes", () => {
    const a = registryWithRole("role-test").compose(new ChunkId("role-test"), "yolo", [], []);
    const r2 = new StandardPromptChunkRegistry();
    r2.register(promptChunk("role-test", "DIFFERENT ROLE CONTENT", "role", "static"));
    const b = r2.compose(new ChunkId("role-test"), "yolo", [], []);
    expect(a.ok && b.ok).toBe(true);
    if (a.ok && b.ok) {
      expect(a.composed.block_1_hash).not.toBe(b.composed.block_1_hash);
    }
  });

  it("recomputeBlockHashes agrees with the cached hashes", () => {
    const res = registryWithRole("role-test").compose(new ChunkId("role-test"), "yolo", [], []);
    expect(res.ok).toBe(true);
    if (res.ok) {
      const [b1, b2] = recomputeBlockHashes(res.composed);
      expect(b1).toBe(res.composed.block_1_hash);
      expect(b2).toBe(res.composed.block_2_hash);
    }
  });

  it("hashContent is deterministic and stable", () => {
    expect(hashContent("hello")).toBe(hashContent("hello"));
    expect(hashContent("hello")).not.toBe(hashContent("Hello"));
  });
});

// ── Rule: validate flags malformed compositions ─────────────────────────────

describe("validate", () => {
  it("flags a PerTurn chunk wrongly placed in the Static block", () => {
    const r = new StandardPromptChunkRegistry();
    const composed = {
      chunks: [
        promptChunk("role-x", "x", "role", "static"),
        modePromptChunk("yolo"),
        // A Budget chunk with Static cache_block — simulates a bug.
        {
          id: new ChunkId("bad-budget"),
          content: "b",
          slot: "budget" as const,
          cache_block: "static" as const,
          hash: 0,
        },
      ],
      block_1_hash: 0,
      block_2_hash: 0,
      rendered: null,
    };
    const errs = r.validate(composed);
    expect(
      errs.some((e) => e.kind === "per_turn_chunk_in_static_block" && e.id.value === "bad-budget"),
    ).toBe(true);
  });

  it("flags more than one Mode chunk", () => {
    const r = new StandardPromptChunkRegistry();
    const composed = {
      chunks: [
        promptChunk("role-x", "x", "role", "static"),
        modePromptChunk("yolo"),
        modePromptChunk("always_ask"),
      ],
      block_1_hash: 0,
      block_2_hash: 0,
      rendered: null,
    };
    const errs = r.validate(composed);
    expect(errs.some((e) => e.kind === "conflicting_mode_chunks")).toBe(true);
  });

  it("flags missing Role slot", () => {
    const r = new StandardPromptChunkRegistry();
    const composed = {
      chunks: [modePromptChunk("yolo")],
      block_1_hash: 0,
      block_2_hash: 0,
      rendered: null,
    };
    const errs = r.validate(composed);
    expect(errs.some((e) => e.kind === "missing_required_slot" && e.slot === "role")).toBe(true);
  });

  it("flags missing Mode slot", () => {
    const r = new StandardPromptChunkRegistry();
    const composed = {
      chunks: [promptChunk("role-x", "x", "role", "static")],
      block_1_hash: 0,
      block_2_hash: 0,
      rendered: null,
    };
    const errs = r.validate(composed);
    expect(errs.some((e) => e.kind === "missing_required_slot" && e.slot === "mode")).toBe(true);
  });
});

// ── get / lookup ────────────────────────────────────────────────────────────

describe("get", () => {
  it("returns registered chunks and undefined for unknown ids", () => {
    const r = registryWithRole("role-x");
    expect(r.get(new ChunkId("role-x"))?.id.value).toBe("role-x");
    expect(r.get(new ChunkId("nope"))).toBeUndefined();
  });
});

// ── Mode helpers ────────────────────────────────────────────────────────────

describe("Mode helpers", () => {
  it("approvalPolicy matches the spec", () => {
    expect(modeApprovalPolicy("always_ask")).toBe("always_ask");
    expect(modeApprovalPolicy("auto_edit")).toBe("auto_explain");
    expect(modeApprovalPolicy("plan")).toBe("plan_only");
    expect(modeApprovalPolicy("safe_auto")).toBe("safe_auto");
    expect(modeApprovalPolicy("yolo")).toBe("none");
  });

  it("defaultToolPhase returns planning only for Plan mode", () => {
    expect(modeDefaultToolPhase("plan")).toBe("planning");
    expect(modeDefaultToolPhase("yolo")).toBe("execution");
    expect(modeDefaultToolPhase("safe_auto")).toBe("execution");
  });

  it("promptChunk ids match the cross-language fixture", () => {
    expect(modePromptChunk("always_ask").id.value).toBe("mode-always-ask");
    expect(modePromptChunk("auto_edit").id.value).toBe("mode-auto-edit");
    expect(modePromptChunk("plan").id.value).toBe("mode-plan");
    expect(modePromptChunk("safe_auto").id.value).toBe("mode-safe-auto");
    expect(modePromptChunk("yolo").id.value).toBe("mode-yolo");
  });
});

// ── ComposedPrompt rendering ────────────────────────────────────────────────

describe("renderComposed", () => {
  it("joins chunk contents with a blank line and caches the result", () => {
    const r = registryWithRole("role-test");
    const res = r.compose(new ChunkId("role-test"), "yolo", [], []);
    expect(res.ok).toBe(true);
    if (res.ok) {
      expect(res.composed.rendered).toBeNull();
      expect(renderedStr(res.composed)).toBe("");
      const rendered = renderComposed(res.composed);
      expect(rendered).toContain("you are a test agent");
      expect(rendered).toContain("Mode: Yolo");
      expect(res.composed.rendered).not.toBeNull();
      expect(renderedStr(res.composed)).toBe(rendered);
    }
  });
});

// ── Standard library bootstraps cleanly ─────────────────────────────────────

describe("standard chunk library", () => {
  it("registers cleanly and produces the canonical coding-agent composition", () => {
    const r = new StandardPromptChunkRegistry();
    expect(r.registerStandardChunks()).toBeNull();
    expect(r.get(new ChunkId("role-coding-agent"))).toBeDefined();
    expect(r.get(new ChunkId("capability-bash"))).toBeDefined();
    expect(r.get(new ChunkId("skill-testing"))).toBeDefined();

    const res = r.compose(
      new ChunkId("role-coding-agent"),
      "safe_auto",
      [
        new ChunkId("capability-bash"),
        new ChunkId("capability-filesystem"),
        new ChunkId("capability-git"),
      ],
      [new ChunkId("skill-testing"), new ChunkId("skill-security-review")],
    );
    expect(res.ok).toBe(true);
    if (res.ok) {
      expect(res.composed.chunks).toHaveLength(1 + 1 + 3 + 2);
      expect(res.composed.chunks[0]?.slot).toBe("role");
      expect(res.composed.chunks[1]?.slot).toBe("mode");
      expect(res.composed.chunks[1]?.id.value).toBe("mode-safe-auto");
    }
  });

  it("includes one chunk per Mode variant", () => {
    const chunks = standardChunks();
    const modeIds = chunks.filter((c) => c.slot === "mode").map((c) => c.id.value);
    expect(modeIds.sort()).toEqual(
      ["mode-always-ask", "mode-auto-edit", "mode-plan", "mode-safe-auto", "mode-yolo"].sort(),
    );
  });
});
